go-controller/pkg/ovn/ovn.go

```diff
@@ -59,20 +57,57 @@ type ACLLoggingLevels struct {
 	Deny  string `json:"deny,omitempty"`
 }
 
-// Controller structure is the object which holds the controls for starting
-// and reacting upon the watched resources (e.g. pods, endpoints)
-type Controller struct {
+// controllerInfo structure is place holder for all fields shared among controllers.
+type controllerInfo struct {
 	client       clientset.Interface
 	kube         kube.Interface
 	watchFactory *factory.WatchFactory
-	stopChan     <-chan struct{}
+	podRecorder  *metrics.PodRecorder
+
+	// event recorder used to post events to k8s
+	recorder record.EventRecorder
+
+	// libovsdb northbound client interface
+	nbClient libovsdbclient.Client
+
+	// libovsdb southbound client interface
+	sbClient libovsdbclient.Client
+
+	// has SCTP support
+	SCTPSupport bool
+}
+
+type Controller interface {
+	Start(ctx context.Context) error
+	Stop()
+
+	GetInternalCacheEntry(eventObjType reflect.Type, obj interface{}) interface{}
+
+	AddResource(eventObjType reflect.Type, obj interface{}, fromRetryLoop bool, extraParameters interface{}) error
+	UpdateResource(eventObjType reflect.Type, oldObj, newObj interface{}, inRetryCache bool, extraParameters interface{}) error
+	DeleteResource(eventObjType reflect.Type, obj, cachedObj, extraParameters interface{}) error
+	GetSyncFunc(reflect.Type) (func([]interface{}) error, error)
+
+	RecordAddEvent(eventObjType reflect.Type, obj interface{})
+	RecordUpdateEvent(eventObjType reflect.Type, obj interface{})
+	RecordDeleteEvent(eventObjType reflect.Type, obj interface{})
+	RecordSuccessEvent(eventObjType reflect.Type, obj interface{})
+	RecordErrorEvent(eventObjType reflect.Type, err error, reason string, obj interface{})
+}
+
```

Comment:
I think we are mixing two interfaces in one. We could have a separate
```
type ManagedNetworkController interface {
	Start(ctx context.Context) error
	Stop()
}
```
in package `controller-manager/controller.go`

Reply:
As for the rest of the interface, I see that it mimics the obj_retry EventHandler interface in what seems a delegation setup from masterEventHandler to this one.
I don't know what to make of it yet. What's the plan? At first sight is difficult to tell because the methods are pretty generic and similar to the EventHandler interface, there is nothing specific to a network controller about them.
So why can't each network controller provide their own implementation of the event handler interface (which could have a base shared implementation) instead of adding this additional interface and delegation layer?

Reply:
because each type of controller has its own add/update/deleteResouce(), for example, secondaryL3Controller only handles nodes and Pods events (and multi-networkpolicy once network-policy is supported). L2controller could only handles pods events. what they do when handling the specific resouce is different (create different ovn logical entities for example).

Reply:
> I think we are mixing two interfaces in one. We could have a separate

Reply:
> because each type of controller has its own add/update/deleteResouce(), for example, secondaryL3Controller only handles nodes and Pods events (and multi-networkpolicy once network-policy is supported). L2controller could only handles pods events. what they do when handling the specific resouce is different (create different ovn logical entities for example).

You can still do that if each controller directly implemented the EventHandler interface instead of adding this new interface.

Reply:
> I don't see benefit of it but only makes code complicated.

The benefit is proper design IMO. The NetworkControllerManager does not care at all about `AddResource`, `UpdateResource`, .... The NetworkControllerManager is only concerned about how to start and stop the network controller and it should own the interface that it needs to achieve this. Likewise, obj_retry_master does not care about starting and stopping the network controllers.

I don't see how it makes the code more complicated. Can you explain?

Reply:
> You can still do that if each controller directly implemented the EventHandler interface instead of adding this new interface.

you mean the current EventHandler takes an extra controller argument? I think It would be clearer to make them as interface of controller as they are different per-controller?

Reply:
> you mean the current EventHandler takes an extra controller argument?

no

>  implement different eventHandler for different controllers?

Yes. 

But a baseEventHandler is not the best option either because if a controller type overrides anything from baseEventHandler then the other baseEventHandler methods won't see it. Golang does not have the typical OOO override.

So the easiest probably is to have shared utility methods for the the stuff from the EventHandler interface that needs a common implementation across the different controller types.

> I think It would be clearer and easier to make them as interface of controller as they are different per-controller?

We already have an EventHandler interface to have events processed in different ways. But what you are doing here is sharing the code of `AreResourcesEqual`, `GetInternalCacheEntry`, `IsResourceScheduled` and `IsObjectInTerminalState` via defining a new interface that adds nothing else and delegating, when there are multiple other alternatives to share this code.

Reply:
If you insist on having an specific interface, it could be

```
type NetworkControllerEventHandler interface {
	GetInternalCacheEntry(eventObjType reflect.Type, obj interface{}) interface{}

	AddResource(eventObjType reflect.Type, obj interface{}, fromRetryLoop bool, extraParameters interface{}) error
	UpdateResource(eventObjType reflect.Type, oldObj, newObj interface{}, inRetryCache bool, extraParameters interface{}) error
	DeleteResource(eventObjType reflect.Type, obj, cachedObj, extraParameters interface{}) error
	GetSyncFunc(reflect.Type) (func([]interface{}) error, error)

	RecordAddEvent(eventObjType reflect.Type, obj interface{})
	RecordUpdateEvent(eventObjType reflect.Type, obj interface{})
	RecordDeleteEvent(eventObjType reflect.Type, obj interface{})
	RecordSuccessEvent(eventObjType reflect.Type, obj interface{})
	RecordErrorEvent(eventObjType reflect.Type, err error, reason string, obj interface{})
}
```

This interface should be in obj_retry_master.go

Reply:
that is what I mean, then we need to have DefaultNetworkControllerEventHandler/SecondaryL3NetworkControllerEventHandler.... Is my understanding correct?

Reply:
the same feeling is for the ManagedNetworkController proposal. It is needs defaultManagedNetworkController/secondaryl3ManagedNetworkController etc, and all needs controller point to do its work.

Reply:
How is it more complicated? In the node side we actually have more than one event handler (node & gateway) and both have a pointer to the controller they are handling events for. So we would just follow the same pattern.

You could have something like this in `defaultcontroller.go`
```
// NewDefaultController creates a new OVN controller for creating logical network
// infrastructure and policy for default l3 network
func NewDefaultController(cc ControllerConnection) *DefaultNetworkController {
	...
        c.initRetryFrameworkForDefaultNetworkController()
        ...
}

func (oc *DefaultNetworkController) initRetryFrameworkForDefaultNetworkController() {
	// Init the retry framework for pods, namespaces, nodes, network policies, egress firewalls,
	// egress IP (and dependent namespaces, pods, nodes), cloud private ip config.
	oc.retryPods = newRetryFrameworkForDefaultNetworkController(oc, oc.watchFactory, factory.PodType)
       ...
}

func (oc *DefaultNetworkController) newRetryFrameworkForDefaultNetworkController(objType objectType, syncFunc func([]interface{}) error, extraParameters interface{}) *retry.RetryFramework {
	eventHandler := &DefaultNetworkControllerEventHandler{
			objType:         objectType,
			watchFactory:    oc.watchFactory,
			oc:              oc,
			extraParameters: extraParameters, // in use by network policy dynamic watchers
			syncFunc:        syncFunc,
		}
       return newRetryFrameworkMasterWithParameters(objType, eventHandler)
}

type DefaultNetworkControllerEventHandler struct {
	masterEvenHandler
}

func (eh *DefaultNetworkControllerEventHandler) AreResourcesEqual(obj1, obj2 interface{}) (bool, error) {
	// implemented in masterEvenHandler os as a free method somewhere
	return eh.areResourcesEqual(obj1, obj2 interface{})
}

func (eh *DefaultNetworkControllerEventHandler) AddResource(obj interface{}, fromRetryLoop bool) error {
	...
}

...
```

Reply:
because for each kind of controller, we'd need to both implment its controller type and the controllerEventHandler type, while lots of their method would be the same (AreResourcesEqual, , IsResourceScheduled, IsObjectInTerminalState) so we may need a baseEventController as well. While in this PR, we'd mostly focus on the difference of each kind of Controller.  I don't see why it is not more complicated. Maybe we can discuss it over meeting next Tuesday?

Reply:
If their method would be same, there is multiple other alternatives to share that code than the one proposed in this PR.

It seems like we have a different perception of what is more complicated. For me, making the code potentially difficult to understand and work with in the future is what makes it more complicated; and not necessarily having more lines of code or implementing new types.

I am trying to reach a compromise here (because there might be other alternatives that would require much more effort). My opinion is that introducing a undocumented interface just to share the implementation of some methods is almost never the correct way to do it. I am proposing an alternative that it is not much more complicated at all, and that we are already doing today in the code. (see https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/obj_retry_node.go vs https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/obj_retry_gateway.go). We don't necessarily need a baseEventController, it could just be utility free methods.

I tried to reach a compromise again and offered an alternative to keep this interface, but in a better place place, with a better name and hopefully some documentation; which overall helps to understand why it is so similar but different than the EventHandler interface. 

I also suggested to split this interface in two because we are mixing two things with different purposes that have nothing to do with each other. I don't know why this is perceived as more complicated

I am doing these suggestions because I think it is worth the tiny bit more of extra effort of having a few more lines of code now to make it easier to read, understand and work with for people in the future; overall trying to help increasing the quality of this PR as well.

Reply:
> the same feeling is for the ManagedNetworkController proposal. It needs defaultManagedNetworkController/secondaryl3ManagedNetworkController etc, and all needs controller pointer to do its work.

I don't think so. The controllers would only need to implement the ManagedNetworkController interface, we don't need new types for this.